packages/graphql/src/services/sessions.ts

```diff
@@ -1569,17 +1569,100 @@ export async function getSessionEvaluation(
     feedbacks: session.feedbacks,
     confusionFeedbacks: session.confusionFeedbacks,
   }
+}
 
-  return {
-    id: `${id}-eval`,
-    status: session.status,
-    isGamificationEnabled: session.isGamificationEnabled,
-    blocks: activeBlock ? [...executedBlocks, activeBlock] : executedBlocks,
-    instanceResults: [
-      ...completeQuestionData(executedInstanceResults),
-      ...completeQuestionData(activeInstanceResults),
-    ],
-    feedbacks: session.feedbacks,
-    confusionFeedbacks: session.confusionFeedbacks,
+export async function cancelSession(
+  { id }: { id: string },
+  ctx: ContextWithUser
+) {
+  const session = await ctx.prisma.session.findUnique({
+    where: { id },
+    include: {
+      activeBlock: true,
+      blocks: {
+        include: {
+          instances: true,
+          leaderboard: true,
+          activeInSession: true,
+        },
+      },
+    },
+  })
+
+  if (!session) return null
+
+  if (session.status !== SessionStatus.RUNNING) {
+    throw new Error('Session is not running')
   }
+
+  const leaderboardEntries = session.blocks
+    .map((block) => block.leaderboard)
+    .flat()
+  const instances = session.blocks.map((block) => block.instances).flat()
+
+  await ctx.prisma.session.update({
+    where: { id },
+    data: {
+      status: SessionStatus.SCHEDULED,
+      startedAt: null,
+      pinCode: null,
+      activeBlock: {
+        disconnect: true,
+      },
+      leaderboard: {
+        deleteMany: {},
+      },
+      feedbacks: {
+        deleteMany: {},
+      },
+      confusionFeedbacks: {
+        deleteMany: {},
+      },
+      blocks: {
+        updateMany: {
+          where: {
+            status: {
+              in: [SessionBlockStatus.EXECUTED, SessionBlockStatus.ACTIVE],
+            },
+          },
+          data: {
+            status: SessionBlockStatus.SCHEDULED,
+            expiresAt: null,
+            execution: 0,
+          },
+        },
+      },
+    },
+  })
+
+  await ctx.prisma.leaderboardEntry.deleteMany({
+    where: {
+      id: {
+        in: leaderboardEntries.map((entry) => entry.id),
+      },
+    },
+  })
+
+  await Promise.all(
+    instances.map(async (instance) => {
+      await ctx.prisma.questionInstance.update({
+        where: {
+          id: instance.id,
+        },
+        data: {
+          participants: 0,
+          // TODO: ensure that this typescript error does not point to a real potential issue
+          results: prepareInitialInstanceResults(instance.questionData),
```

Comment:
This line create TypeScript errors at a potentially critical point. If this object is corrupted and code in other components relies on the default structure, this could lead to significant problems...

Reply:
One TypeScript error originating from the initialization of the results object will be fixed later